docs/internal_abi.md

```diff
@@ -0,0 +1,323 @@
+The remaining type (`U64`) requires 64 bits for payload and 0 bits for discriminator, and so does not fit into the significand part of a `Float64` signalling NaN.
+The first special case is therefore not used
+The second special case is not used because neither type is a pointer.
+
+We are now left requiring one bit of discriminator and 64 bits of payload.
+The payload size is a power of two, and so the address of the union type is the address of the start of a 64-bit payload.
+The 1-bit discriminator is stored afterwards, with a value of 0 representing a `Float64` and a value of `1` representing a `U64`.
+The field-packing algorithm is free to use the remaining 7 bits in the discriminator for other values.
+
+### Example: `Float32 | U20 | U16`
+
+The size of `Float32` is 32 bits and requires one discriminator state.
+The size of `U20` is 20 bits and requires one discriminator state.
+The size of `U16` is 16 bits and requires one discriminator state.
+
+The first special case applies: the largest type is a floating-point type and the remaining types require 20 bits of payload and one of discriminator, which fits in the significand for a `Float32` signalling NaN.
+
+Placing the discriminator in the low bits would give one byte and two partial bytes for the payload.
+Placing the discriminator in the high bits would give two complete bytes and one nybble for the payload.
+The disciminator is therefore placed in bit 20, leaving bits 0-19 for the payload.
+
+The `U20` uses all of the available payload bits and so is stored in bits 0-19.
+The `U16` follows the rule that it should be aligned to the end of the opposite end to the discriminator and is therefore stored in bits 0-15.
+
+A match expression first loads the entire 32-bit value and compares bits 21-31 against an all-ones bit pattern.
+If the comparison result is not-equal, the entire value is a `Float32`.
+If the comparison result is equal, then a value of bit 20 is tested, with zero indicating a `U16` and one indicating a `U20`, either of which can be extracted by a bit shift (note: On Arm systems, either value can be extracted with a single `UBFX` instruction, on x86 systems that support the BMI1 instructions a BEXTR can do the same).
+
+### Example: `Float64 | T* | U*`
+
+The payload size of `Float64` is 64 bits and it requires one discriminator state.
+The payload size of `T*` and `U*` is each either 32, 48, or 128 bits, depending on the target architecture and they require one discriminator state between them.
+The lowering for this union therefore depends on the target.
+
+On 32-bit systems, the largest type is `Float64` and the remaining types all fit easily in the significand bits.
+There are no discriminator bits required but the 'as-if' rules triggers as if there were a single bit for determining which end of the space should be used and the remaining values are stored at the end of the space that is strongly aligned: bits 0-31.
```

Comment:
Shouldn't there be a discriminator for the difference between `U*` and `T*`?

Reply:
No, you can always do the equivalent of `((Verona::Object*)ptr)->descriptor` and compare that against the descriptor pointers for the matched types for any concrete type.  @sylvanc, I'm not 100% sure what you do when you have something like `Cls1 | Cls2 | Interface1` and want to match on `Interface1`?